# analysing comments in Hacker News website
we will analyse two types of posts, Ask HN and Show HN, Ask HN are those pot that asks questions.  Show HN are some general posts that are interesting.

We'll analyse which type of posts get more comments on average. and is there certain of the day when more comments are received. 

In [1]:
from csv import reader
open_file = open("hacker_news.csv")
read_file = reader(open_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[:1]
hn = hn[1:]
print(headers)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts),len(show_posts), len(other_posts))

1744 1162 17194


In [4]:
total_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments/len(ask_posts)

print("Average Ask comments", avg_ask_comments)

total_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments/len(show_posts)

print("Average Show comments", avg_show_comments)

Average Ask comments 14.038417431192661
Average Show comments 10.31669535283993


We found that Ask Comments receive about 4 more comments on average than show comments

In [5]:
import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append(
    [row[6], int(row[4])] #index 6 is created at, index 4 is comment number
    )

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = row[0]
    comment = row[1]
    comment_hour = dt.datetime.strptime(date, "%m/%d/%Y %H:%M").strftime("%H")
    
    if comment_hour in counts_by_hour:
        counts_by_hour[comment_hour] += 1
        comments_by_hour[comment_hour] += comment
    else:
        counts_by_hour[comment_hour] = 1
        comments_by_hour[comment_hour] = comment

print(counts_by_hour, "\n" )
print(comments_by_hour)



{'06': 44, '07': 34, '02': 58, '18': 109, '13': 85, '08': 48, '05': 46, '04': 47, '22': 71, '20': 80, '23': 68, '15': 116, '01': 60, '16': 108, '19': 110, '00': 55, '11': 58, '10': 59, '21': 109, '14': 107, '09': 45, '17': 100, '12': 73, '03': 54} 

{'06': 397, '07': 267, '02': 1381, '18': 1439, '13': 1253, '08': 492, '05': 464, '04': 337, '22': 479, '20': 1722, '23': 543, '15': 4477, '01': 683, '16': 1814, '19': 1188, '00': 447, '11': 641, '10': 793, '21': 1745, '14': 1416, '09': 251, '17': 1146, '12': 687, '03': 421}


In [6]:
avg_by_hour = []

for row in counts_by_hour:
    hour = row
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])
    
avg_by_hour

[['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['13', 14.741176470588234],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['20', 21.525],
 ['23', 7.985294117647059],
 ['15', 38.5948275862069],
 ['01', 11.383333333333333],
 ['16', 16.796296296296298],
 ['19', 10.8],
 ['00', 8.127272727272727],
 ['11', 11.051724137931034],
 ['10', 13.440677966101696],
 ['21', 16.009174311926607],
 ['14', 13.233644859813085],
 ['09', 5.5777777777777775],
 ['17', 11.46],
 ['12', 9.41095890410959],
 ['03', 7.796296296296297]]

In [7]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[9.022727272727273, '06'], [7.852941176470588, '07'], [23.810344827586206, '02'], [13.20183486238532, '18'], [14.741176470588234, '13'], [10.25, '08'], [10.08695652173913, '05'], [7.170212765957447, '04'], [6.746478873239437, '22'], [21.525, '20'], [7.985294117647059, '23'], [38.5948275862069, '15'], [11.383333333333333, '01'], [16.796296296296298, '16'], [10.8, '19'], [8.127272727272727, '00'], [11.051724137931034, '11'], [13.440677966101696, '10'], [16.009174311926607, '21'], [13.233644859813085, '14'], [5.5777777777777775, '09'], [11.46, '17'], [9.41095890410959, '12'], [7.796296296296297, '03']]


In [8]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [13]:
print("Top 5 Hours for 'Ask HN' Comments")
for avg, hour in sorted_swap[:5]:
    hour_f = dt.datetime.strptime(hour, "%H")
    print(
        "{}: {:.2f}  average comments per post".format(
    hour_f.strftime("%H:%M"), avg)
        )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59  average comments per post
02:00: 23.81  average comments per post
20:00: 21.52  average comments per post
16:00: 16.80  average comments per post
21:00: 16.01  average comments per post


# conclusion

From the analysis above we can see the best time to create a post is at 3:00 PM because it receives a on avg of 38.59 comments. the next best time is 2:00 AM which receives 23.81 comments on average. 